src/box/txn.c

```diff
@@ -340,27 +340,48 @@ txn_stmt_prepare_rollback_info(struct txn_stmt *stmt, struct tuple *old_tuple,
 		tuple_ref(stmt->rollback_info.new_tuple);
 }
 
+/**
+ * Run the statement's `on_rollback` triggers, if any.
+ */
+static void
+txn_stmt_run_on_rollback_triggers(struct txn_stmt *stmt)
+{
+	if (stmt->has_triggers && trigger_run(&stmt->on_rollback, stmt) != 0) {
+		diag_log();
+		panic("statement rollback trigger failed");
+	}
+}
+
 /*
  * Undo changes done by a statement and run the corresponding
  * rollback triggers.
  *
+ * Logically, we call triggers after running statements. These triggers can
+ * make significant changes (for instance, DDL triggers), so, for consistency,
+ * we should call the statement's `on_rollback` triggers before rolling back
+ * the corresponding statement.
+ *
  * Note, a trigger set by a particular statement must be run right
- * after the statement is rolled back, because rollback triggers
+ * before the statement is rolled back, because rollback triggers
  * installed by DDL statements restore the schema cache, which is
  * necessary to roll back previous statements. For example, to roll
  * back a DML statement applied to a space whose index is dropped
  * later in the same transaction, we must restore the dropped index
  * first.
+ *
+ * Note, the _priv space rollback trigger accesses the database from
+ * `user_reload_privs` (gh-xxx), so it needs to be run after rolling back the
```

Comment:
If you agree with this patch and that the `_priv` space rollback trigger works wrong, I will open a ticket.